### PR TITLE
Custom typescript config

### PR DESCRIPTION
**Pull request by keichinger**

### CODE REVIEW

File: lib/Kaba.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 const kabaBabelPreset = require("kaba-babel-preset");
 const kleur = require("kleur");
+const fs = require("fs");
```
Comment (apfelbox) [inline lib/Kaba.js]: Use `fs-extra`

File: lib/Kaba.js [hunk 2]
```diff
@@ -18,6 +19,7 @@ const typeScriptErrorFormatter = require("@becklyn/typescript-error-formatter");
  *      js: {
  *          javaScriptDependenciesFileName: string,
  *          webpack: webpack.Configuration,
+ *          ?tsConfigFile: string,
```
Comment (apfelbox) [inline lib/Kaba.js]: I would name this parameter like your property.

Also: is this valid syntax in JSDoc?

File: lib/Kaba.js [hunk 3]
```diff
@@ -395,13 +403,20 @@ class Kaba
...
             sass: this.buildSassConfig(),
             js: {
                 javaScriptDependenciesFileName: this.javaScriptDependenciesFileName,
                 webpack: this.buildWebpackConfig(cliConfig),
             },
         };
+
+        if (null != this.customTypeScriptConfig)
+        {
+            buildConfig.js.tsConfigFile = this.customTypeScriptConfig;
```
Comment (apfelbox) [inline lib/Kaba.js]: Just always append it.

File: lib/Kaba.js [hunk 4]
```diff
@@ -547,6 +562,15 @@ class Kaba
...
 
+            let tsConfig = path.join(this.libRoot, "tsconfig.json");
+            let customTsConfig = path.join(this.cwd, "tsconfig.json");
+
+            if (fs.existsSync(customTsConfig))
+            {
+                tsConfig = customTsConfig;
+                this.customTypeScriptConfig = customTsConfig;
+            }
+
```
Comment (apfelbox) [inline lib/Kaba.js]:
```js
let customTypeScriptConfig = path.join(this.cwd, "tsconfig.json");

if (fs.existsSync(customTypeScriptConfig))
{
    this.customTypeScriptConfig = customTypeScriptConfig;
}
```

File: lib/Kaba.js [hunk 5]
```diff
@@ -557,7 +581,7 @@ class Kaba
                         loader: "ts-loader",
                         options: {
                             context: this.cwd,
-                            configFile: path.join(this.libRoot, "tsconfig.json"),
+                            configFile: tsConfig,
```
Comment (apfelbox) [inline lib/Kaba.js]:
```js
configFile: this.customTypeScriptConfig || path.join(this.libRoot, "tsconfig.json");
```

File: lib/runner/WebpackRunner.js [hunk 6]
```diff
@@ -73,6 +73,13 @@ class WebpackRunner
 
                 const compiler = webpack(this.webpackConfig);
 
+                if (null != this.buildConfig.js.tsConfigFile)
+                {
+                    let truncatedPath = this.buildConfig.js.tsConfigFile.replace(process.cwd(), ".");
+
+                    this.logger.log(`Using custom TypeScript config: ${kleur.yellow(truncatedPath)}`);
```
Comment (apfelbox) [inline lib/runner/WebpackRunner.js]:
```js
this.logger.log(`Using custom TypeScript config: ${kleur.yellow(path.relative(cwd, this.buildConfig.js.tsConfigFile)}`);
```

You need to pass `cwd` via the build config as additional parameter.

File: bin/run.js [hunk 7]
```diff
@@ -129,12 +129,14 @@ function runKaba (opts, isVerbose)
         logger.log("kaba started");
         const start = process.hrtime();
         const cliConfig = new CliConfig(opts);
+        let workingDirectory = process.cwd();
```
Comment (apfelbox) [inline bin/run.js]: Kaba already has a `cwd`: https://github.com/Becklyn/kaba/blob/master/lib/Kaba.js#L39

Please use this exact value and pass it out through the build config, as mentioned in https://github.com/Becklyn/kaba/pull/50#discussion_r256862775

File: lib/Kaba.js [hunk 8]
```diff
@@ -403,20 +403,14 @@ class Kaba
      */
     getBuildConfig (cliConfig)
     {
-        let buildConfig = {
+        return {
             sass: this.buildSassConfig(),
             js: {
                 javaScriptDependenciesFileName: this.javaScriptDependenciesFileName,
                 webpack: this.buildWebpackConfig(cliConfig),
+                customTypeScriptConfig: this.customTypeScriptConfig,
```
Comment (apfelbox) [inline lib/Kaba.js]: The parameter definition for this field is missing, so you are currently creating a dynamic class which is generally not a good idea.

Please add it and initialize it to `null`. You need to adjust your type above, then.

File: bin/run.js (hunk 7 above)
Comment (keichinger) [reply bin/run.js]: Well, I was just re-using existing code from this file :)

I've added a getter in `Kaba` to access its `cwd`.

**Pull request merged**